### PR TITLE
DOCS-1083: MinIO Server Release RELEASE.2023-12-02T10-51-33Z

### LINKED ISSUE #1083: [RELEASE] MinIO Server RELEASE.2023-12-02T10-51-33Z
MinIO Server [RELEASE.2023-12-02T10-51-33Z](https://github.com/minio/minio/releases/tag/RELEASE.2023-12-02T10-51-33Z) has changes that impact the docs:

- [ ] Adds Websocket Grid functionality (but maybe not in the main docs; more for internal reference for now) ([PR #18461](https://github.com/minio/minio/pull/18461))
- [x] Adds snowball functionality to batch replicate jobs ([PR #18485](https://github.com/minio/minio/pull/18485))
   - [x] Related: adds ability for immediate tiering for snowball replication ([PR #18508](https://github.com/minio/minio/pull/18508))
- [ ] Adds ability to use external cache for GET/HEAD functions (for HDD drives primarily) ([PR #18239](https://github.com/minio/minio/pull/18239))
- [x] Relaxes batch limits for Kafka events ([PR #18513](https://github.com/minio/minio/pull/18513))
- [ ] Adds a drive max timeout ([PR #18501](https://github.com/minio/minio/pull/18501))
- [x] Adds support for an age conditional for bucket policies for pre-signed URLs ([PR #18529](https://github.com/minio/minio/pull/18529))
- [x] Additional Kafka variables for compression ([PR #18562](https://github.com/minio/minio/pull/18562))
- [x] Adds new expiration job type to the batch job framework ([PR #17946](https://github.com/minio/minio/pull/17946))

### PR DESCRIPTION
**Pull request by ravindk89**
# Summary

Closes #1083 

Also reorganizes the batch framework pages. Could use additional refinement, but considering that out of scope for now.

Staged:

- http://192.241.195.202:9000/staging/DOCS-1083/linux/index.html
- http://192.241.195.202:9000/staging/DOCS-1083/linux/administration/batch-framework.html
- http://192.241.195.202:9000/staging/DOCS-1083/linux/administration/batch-framework-job-replicate.html#batch-compression
- http://192.241.195.202:9000/staging/DOCS-1083/linux/administration/batch-framework-job-replicate.html#batch-compression (``snowball``)
- http://192.241.195.202:9000/staging/DOCS-1083/linux/administration/identity-access-management/policy-based-access-control.html#supported-s3-policy-condition-keys
- http://192.241.195.202:9000/staging/DOCS-1083/linux/reference/minio-server/settings/notifications/kafka.html#compression-codec
- http://192.241.195.202:9000/staging/DOCS-1083/linux/reference/minio-server/settings/notifications/kafka.html#batch-size

### CODE REVIEW

File: source/administration/batch-framework-job-replicate.rst [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+         | - Specify ``off`` for Virtual style
+         | - Specify ``auto`` to let MinIO determine the correct lookup style.
+         |
+         | Defaults to ``auto``.
+
+     * - ``credentials:`` 
+       - | The ``accesskey:`` and ``secretKey:`` or the ``sessionToken:`` that grants access to the object(s).
+         | Only specify for the deployment that is not the :ref:`local <minio-batch-local>` deployment. 
+
+     * - ``snowball``
```
Comment (ravindk89) [inline source/administration/batch-framework-job-replicate.rst:141]: @vadmeste

File: source/administration/batch-framework-job-replicate.rst [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+The access to objects and buckets for the replication process may be restricted by the credentials you provide in the YAML for either the source or target destinations. 
+
+.. versionchanged:: MinIO Server RELEASE.2023-04-07T05-28-58Z
+
+   You can replicate from a remote MinIO deployment to the local deployment that runs the batch job.
+
+For example, you can use a batch job to perform a one-time replication sync to push objects from a bucket on a local deployment at ``minio-local/invoices/`` to a bucket on a remote deployment at ``minio-remote/invoices``.
+You can also pull objects from the remote deployment at ``minio-remote/invoices`` to the local deployment at ``minio-local/invoices``.
+
+Batch Compression
```
Comment (ravindk89) [inline source/administration/batch-framework-job-replicate.rst:77]: @vadmeste

File: source/administration/batch-framework-job-expire.rst [hunk 3]
```diff
@@ -0,0 +1,179 @@
+.. _minio-batch-framework-expire-job:
+
+=================
+Batch Expiration
```
Comment (ravindk89) [inline source/administration/batch-framework-job-expire.rst:4]: @krisis

**Review by ravindk89 — COMMENTED** [review]
Quick ping on relevant sections for engineering.

File: source/administration/batch-framework-job-expire.rst [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+   * - ``apiVersion``
+     - *Required*
+        
+       Set to ``v1``.
+
+   * - ``bucket``
+     - *Required*
+       
+       Specify the name of the object in which the job runs.
```
[the remark this answers is not in the document]
Comment (djwfyi) [reply source/administration/batch-framework-job-expire.rst:64]: We have one YAML file per bucket?
Or one expire batch job type section per bucket, I guess?

File: source/administration/batch-framework-job-expire.rst [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+   * - ``prefix``
+     - *Optional*
+      
+       Specify the bucket prefix in which the job runs.
+
+   * - ``rules``
+     - *Required*
+       
+       An array of one or more expiration rules to apply to objects in the specified ``bucket`` and ``prefix`` (if any).
```
Comment (djwfyi) [inline source/administration/batch-framework-job-expire.rst:74]: Is this rule IDs? If so, might be useful to link to something that says how they can find those.
Or does the array have a rule configuration string?

File: source/administration/batch-framework-job-expire.rst [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+       Supports glob-style wildcards (``*``, ``?``).
+
+   * - ``rules.[n].olderThan``
+     - *Optional*
+
+       Specify the age of objects for filtering objects.
+       The rule applies to only those objects older than the specified unit of time.
+
+       Supports values such as ``72h`` or ``3d`` for objects three days old.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-expire.rst:97]:
```suggestion
       For example, ``72h`` or ``3d`` selects objects older than three days.
```

File: source/administration/batch-framework-job-expire.rst [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+       Specify an ISO-8601 timestamp for filtering objects.
+
+       The rule applies to only those objects created *before* the specified timestamp.
+
+   * - ``rules.[n].tags``
+     - *Optional*
+
+       Specify an array of key-value pairs describing object tags to use for filtering objects.
+       The ``value`` key supports glob-style wildcards (``*``, ``?``).
```
Comment (djwfyi) [inline source/administration/batch-framework-job-expire.rst:110]:
```suggestion
       The ``value`` entry supports glob-style wildcards (``*``, ``?``).
```

File: source/administration/batch-framework-job-expire.rst [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+   * - ``notify.endpoint``
+     - *Optional*
+
+       The predefined endpoint to send events for notifications.
+
+   * - ``notify.token``
+     - *Optional*
+
+       An optional :abbr:`JWT <JSON Web Token>` to access the ``notify.endpoint``.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-expire.rst:161]: abbr isn't rendering in the list.

```suggestion
       An optional JSON Web Token (JWT) to access the ``notify.endpoint``.
```

File: source/administration/batch-framework-job-keyrotate.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+   :depth: 2
+
+.. versionadded:: MinIO RELEASE.2023-04-07T05-28-58Z
+
+The MinIO Batch Framework allows you to create, manage, monitor, and execute jobs using a YAML-formatted job definition file (a "batch file").
+The batch jobs run directly on the MinIO deployment to take advantage of the server-side processing power without constraints of the local machine where you run the :ref:`MinIO Client <minio-client>`.
+
+The ``keyrotate`` batch job type cycles the :ref:`sse-s3 or sse-kms keys <minio-sse-data-encryption>` for encrypted objects on a MinIO deployment.
+
+The YAML configuration supports filters to restrict key rotation to  a specific set of objects by creation date, tags, metadata, or kms key.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-keyrotate.rst:21]:
```suggestion
The YAML configuration supports filters to restrict key rotation to a specific set of objects by creation date, tags, metadata, or kms key.
```

File: source/administration/batch-framework-job-keyrotate.rst [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+For **notifications**
+
+.. list-table::
+   :widths: 25 75
+   :width: 100%
+
+   * - ``endpoint:`` 
+     - The predefined endpoint to send events for notifications.
+   * - ``token:`` 
+     - An optional :abbr:`JWT <JSON Web Token>` to access the ``endpoint``.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-keyrotate.rst:93]:
```suggestion
     - An optional JSON Web Token (JWT) to access the ``endpoint``.
```

File: source/administration/batch-framework-job-replicate.rst [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+The remote MinIO deployment can check and immediately apply lifecycle management tiering rules to batched objects.
+You can modify the compression settings in the :ref:`replicate <minio-batch-job-types>` job configuration.
+
+.. _minio-batch-framework-replicate-job-ref:
+
+Replicate Batch Job Reference
+-----------------------------
+
+The YAML **must** define the source and target deployments.
+If the _source_ deployment is remote, then the _target_ deployment **must** be ``local``.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-replicate.rst:91]:
```suggestion
If the *source* deployment is remote, then the *target* deployment **must** be ``local``.
```

File: source/administration/batch-framework-job-replicate.rst [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+         | - Specify ``auto`` to let MinIO determine the correct lookup style.
+         |
+         | Defaults to ``auto``.
+
+     * - ``credentials:`` 
+       - | The ``accesskey:`` and ``secretKey:`` or the ``sessionToken:`` that grants access to the object(s).
+         | Only specify for the deployment that is not the :ref:`local <minio-batch-local>` deployment. 
+
+     * - ``snowball``
+       - | .. versionadded:: RELEASE.2023-12-09T18-17-51Z
```
Comment (djwfyi) [inline source/administration/batch-framework-job-replicate.rst:142]:
```suggestion
       - | *version added*: RELEASE.2023-12-09T18-17-51Z
```

Comment (djwfyi) [reply source/administration/batch-framework-job-replicate.rst:142]: Formatting isn't working inside the list.

File: source/administration/batch-framework-job-replicate.rst [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+     * - ``snowball.batch``
+       - | Specify the maximum integer number of objects to batch for compression.
+         | Defaults to ``100``.
+
+     * - ``snowball.inmemory``
+       - | Specify ``false`` to stage archives using local storage or ``true`` to stage to memory (RAM).
+         | Defaults to ``true``.
+
+     * - ``snowball.compress``
+       - | Specify ``true`` to generate archives using the S2/Snappy compression algorithm.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-replicate.rst:159]:
```suggestion
       - | Specify ``true`` to generate archives using the `S2/Snappy compression algorithm <https://en.wikipedia.org/wiki/Snappy_(compression)>`__.
```

File: source/administration/batch-framework-job-replicate.rst [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+         | Defaults to ``false`` or no compression.
+
+     * - ``snowball.smallerThan``
+       - | Specify the size of object in Megabits (MiB) under which MinIO should batch objects.
+         | Defaults to ``5MiB``.
+
+     * - ``snowball.skipErrs``
+       - | Specify ``false`` to direct MinIO to halt on any object which produce errors on read.
```
Comment (djwfyi) [inline source/administration/batch-framework-job-replicate.rst:167]:
```suggestion
       - | Specify ``false`` to direct MinIO to halt on any object which produces errors on read.
```

File: source/administration/batch-framework.rst [hunk 15]
```diff
@@ -76,16 +77,16 @@ The :mc:`mc batch` commands include
...
           :end-before: end-mc-batch-describe-desc
 
+.. _minio-batch-framework-access:
+
 Access to ``mc batch``
 ----------------------
 
-A user's access keys and policies do not restrict the the buckets, prefixes, or objects the batch function can access or the types of actions the process can perform on any objects.
+Each batch job executes using the credentials specified in the batch definition.
+The success of a given batch job depends on those credentials having the appropraite :ref:`permissions <minio-policy>` to perform all requested actions.
```
Comment (djwfyi) [inline source/administration/batch-framework.rst:86]:
```suggestion
The success of a given batch job depends on those credentials having the appropriate :ref:`permissions <minio-policy>` to perform all requested actions.
```

File: source/reference/minio-mc/mc-batch-generate.rst [hunk 16]
```diff
@@ -74,8 +74,11 @@ Parameters
    
    The type of job to generate a YAML document for.
    
-   Currently, :mc:`mc batch` supports the ``replicate`` and ``keyrotate`` job types.
-
+   Supports the following values:
+   
+   -  ``replicate``
```
Comment (djwfyi) [inline source/reference/minio-mc/mc-batch-generate.rst:79]: Should these be links?
Either to the sections below or the administration/batch-replicate docs?

I think surfacing links to the description in administration somewhere other than the bottom of each YAML is worth doing.

File: source/reference/minio-server/settings/notifications/kafka.rst [hunk 17]
```diff
@@ -441,4 +465,59 @@ Comment
...
+
+   .. tab-item:: Configuration Setting
+      :sync: config
+
+      .. mc-conf:: notify_kafka compression_level
+         :delimiter: " "
+
+Controls the level of compression applied based on the configured compression codec.
+
+Specify an integer value greater than or equal to 0.
```
Comment (djwfyi) [inline source/reference/minio-server/settings/notifications/kafka.rst:522]:
```suggestion
Specify an integer value greater than or equal to ``0``.
```

**Review by djwfyi — APPROVED** [review]
Some formatting or wording suggestions to consider.
Maybe surface links from mc-batch-replicate to the administration/batch-replicate docs somewhere.

Otherwise, this LGTM. Don't feel the need to see it again after you address/consider my suggestions.

File: source/administration/batch-framework-job-expire.rst (hunk 8 above)
Comment (ravindk89) [reply source/administration/batch-framework-job-expire.rst]: Interesting!

File: source/reference/minio-mc/mc-batch-generate.rst (hunk 16 above)
Comment (ravindk89) [reply source/reference/minio-mc/mc-batch-generate.rst:79]: :thinking: yeah I can add some reflinks to make this work.

File: source/administration/batch-framework-job-replicate.rst (hunk 2 above)
Comment (vadmeste) [reply source/administration/batch-framework-job-replicate.rst:77]: Maybe we should call this, 'small files replication optimization' ?

File: source/administration/batch-framework-job-replicate.rst [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+     * - ``snowball.batch``
+       - | Specify the maximum integer number of objects to batch for compression.
+         | Defaults to ``100``.
+
+     * - ``snowball.inmemory``
+       - | Specify ``false`` to stage archives using local storage or ``true`` to stage to memory (RAM).
+         | Defaults to ``true``.
+
+     * - ``snowball.compress``
+       - | Specify ``true`` to generate archives using the `S2/Snappy compression algorithm <https://en.wikipedia.org/wiki/Snappy_(compression)>`__.
```
Comment (vadmeste) [inline source/administration/batch-framework-job-replicate.rst:159]: Maybe compress objects on the wire

**Review by vadmeste — COMMENTED** [review]
Few comments

File: source/administration/batch-framework-job-expire.rst [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+       
+       An array of one or more expiration rules to apply to objects in the specified ``bucket`` and ``prefix`` (if any).
+
+   * - ``rules.[n].type``
+     - *Required*
+
+       Supports one of the following two values:
+
+       - ``object`` - apply the expiration to objects and their versions
+       - ``deleted`` - apply the expiration to object Delete Markers
```
Comment (krisis) [inline source/administration/batch-framework-job-expire.rst:82]: This rule applies to a delete marker (current version) and all its versions. I'm not sure if the current description conveys this. We don't want users assume that in the case of a delete marker current version, this rule expires only the (topmost) delete marker and leaves the other (noncurrent) versions as is.

**Review by krisis — COMMENTED** [review]
Great work on Batch Expire feature documentation!

File: source/administration/batch-framework-job-expire.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (krisis) [reply source/administration/batch-framework-job-expire.rst]: Batch jobs apply to one bucket. A batch job yaml must contain the bucket on which it operates.

File: source/administration/batch-framework-job-expire.rst (hunk 19 above)
Comment (ravindk89) [reply source/administration/batch-framework-job-expire.rst:82]: So it's more like

`object ` - deletes only objects and previous versions, ignoring delete markers
`deleted` - implies `object` plus delete markers

File: source/administration/batch-framework-job-expire.rst [hunk 20]
```diff
@@ -78,8 +84,8 @@ Expire Batch Job Reference
 
        Supports one of the following two values:
 
-       - ``object`` - apply the expiration to objects and their versions
-       - ``deleted`` - apply the expiration to object Delete Markers
+       - ``object`` - apply the expiration to an object and it's versions, *ignoring* deleted objects (``DeleteMarker`` is current version).
```
Comment (djwfyi) [inline source/administration/batch-framework-job-expire.rst:87]:
```suggestion
       - ``object`` - apply the expiration to an object and its versions, *ignoring* deleted objects (``DeleteMarker`` is current version).
```

File: source/administration/batch-framework-job-expire.rst [hunk 21]
```diff
@@ -78,8 +84,8 @@ Expire Batch Job Reference
...
+       - ``deleted`` - apply the expiration to an object and it's versions, including deleted objects (``DeleteMarker`` is current version).
```
Comment (djwfyi) [inline source/administration/batch-framework-job-expire.rst:88]:
```suggestion
       - ``deleted`` - apply the expiration to an object and its versions, including deleted objects (``DeleteMarker`` is current version).
```

**Review by djwfyi — COMMENTED** [review]
Swap `it's` for `its` in two spots.

File: source/administration/batch-framework-job-expire.rst [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+       
+       An array of one or more expiration rules to apply to objects in the specified ``bucket`` and ``prefix`` (if any).
+
+   * - ``rules.[n].type``
+     - *Required*
+
+       Supports one of the following two values:
+
+       - ``object`` - apply the expiration to an object and its versions, *ignoring* deleted objects (``DeleteMarker`` is current version).
+       - ``deleted`` - apply the expiration to an object and its versions, including deleted objects (``DeleteMarker`` is current version).
```
Comment (krisis) [inline source/administration/batch-framework-job-expire.rst:88]: `deleted` rule type applies only to objects with delete marker as their current version. i.e an object that won't be visible when listed using ListObjects API.

File: source/administration/batch-framework-job-expire.rst [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+     - *Required*
+       
+       An array of one or more expiration rules to apply to objects in the specified ``bucket`` and ``prefix`` (if any).
+
+   * - ``rules.[n].type``
+     - *Required*
+
+       Supports one of the following two values:
+
+       - ``object`` - apply the expiration to an object and its versions, *ignoring* deleted objects (``DeleteMarker`` is current version).
```
Comment (krisis) [inline source/administration/batch-framework-job-expire.rst:87]: `object` rule type applies objects which don't have delete marker as their current version. i.e a regular object which will appear when its bucket is listed using ListObjects API.

**Review by krisis — CHANGES_REQUESTED** [review]

**Comment by ravindk89** [comment]
@krisis changes updating shortly, staging also updated.

**Review by krisis — APPROVED** [review]
Batch expiration feature docs look good to me.

**Pull request merged**